### ISSUE TITLE
"salt" state.sls 17x slower than "salt-call"

### ISSUE THREAD

**Issue by FraaJad** [saltstack/salt#39998]
### Description of Issue/Question
When running salt command from the master it takes 5m10s to complete the no-op run.  If I run salt-call from the minion, it takes 18s. 

### Setup
The total size of the vhosts which are deployed entirely from pillar is 378KB.
I'm using PillarStack and Consul ext_pillar providers
[mf_llpweb.zip](https://github.com/saltstack/salt/files/839821/mf_llpweb.zip)

### Steps to Reproduce Issue
I have a state that uses nginx.ng from https://github.com/saltstack-formulas/nginx-formula to deploy 25 vhosts.  If the state has already been ran and the current run results in a no-op, if I run "salt-call state.sls roles.mf_llpweb" it completes in 18 seconds.  If I SSH to the saltmaster and run "salt 'servername' state.sls roles.mf_llpweb" it takes 5 minutes and 10 seconds to complete the no-op run.

I tailed the minion log during both runs and there is no particular slow phase.  When running using "salt" command, each log entry seems to take the same long amount of time.   

### Versions Report
I've tried both 2016.11.1 and 2016.11.3

**Comment by ninja-** [comment]
What's the latency between your master and minion

**Comment by ninja-** [comment]
The ping to be exact

**Comment by ninja-** [comment]
Also is the 5 minutes an install on clean machine or an "incremental" run?

**Comment by dougofthemoment** [comment]
My ping from master to minion is 0.340 ms.  This is an incremental run that results in a no-op (103 states succeed with no changes).

**Comment by ninja-** [comment]
That's weird. Could you attach master and minion logs in debug mode?

**Comment by dougofthemoment** [comment]
There is tons of client names and sensitive stuff in there, is there a private place I could drop them?

**Comment by ninja-** [comment]
If that's OK you can put them in some cloud and drop me a link at xninja@openmailbox.org

Disclaimer: I am just a contributor, but I am experienced with solving salt problems similar to yours.

**Comment by Ch3LL** [comment]
@dougofthemoment you might be running into https://github.com/saltstack/salt/issues/38758 which as been fixed in https://github.com/saltstack/salt/pull/39505 can you give that a try?

**Comment by ninja-** [comment]
Yeah @dougofthemoment please try running this on develop branch as some important fixes were not backported

**Comment by ninja-** [comment]
Well OK so that's still not in .3 :)

**Comment by Ch3LL** [comment]
Yeah correct. Just wanted to clarify in case we did need to backport something back from develop. Thanks

**Comment by dougofthemoment** [comment]
I figured it out...  First I did install the develop branch and it speeded it up a bit but it still took minutes.  I dug into the logs and found that an unless statement attached to archive.extracted was failing when it shouldn't which caused it to have to recompute the hash and the enforce_ownership_on option for each dir.  

What is happening is the unless statement is ran using *bash* when executed from "salt-call" and ran using *sh* when executed with "salt".  After I added "shell: /bin/bash" to the state with the unless argument, it is full speed when executed with with "salt".
